### PR TITLE
agent: Scaling logic refactor

### PR DESCRIPTION
**Pull request by sharnoff**
Broadly, the design here is to move to a single state object per-VM (`core.State`) that exposes a _pure_ method, `(*core.State).NextAction()`, which tells the caller what they should do next.

This should allow:

- Unit tests for our scaling logic
- Easier implementation of changes to the scaling flow that require adding/modifying the state we track

In particular, it'll be _much_ easier to address #350 and #23 after this change.

Steps:

- [x] Add `core.State`
- [x] Manage `core.State` from pkg/agent/runner.go, execute `Action`s.
- [x] Add tests for `core.State`

### CODE REVIEW

**Review by tychoish — COMMENTED** [review]
can't wait to see the test ;)

File: pkg/agent/core/core.go [hunk 1]
```diff
@@ -0,0 +1,723 @@
...
+}
+
+type neonvmState struct {
+	lastSuccess *api.Resources
+	// ongoingRequested, if not nil, gives the resources requested
+	ongoingRequested *api.Resources
+	requestFailedAt  *time.Time
+}
+
+func NewState(vm api.VmInfo, config Config, notifyUpdates util.CondChannelSender) *State {
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply pkg/agent/core/core.go:140]: > is this a move from the existing agent package?

not quite sure what you mean?

> `core` is maybe a confusing name?

I'll see if I can think of something better.

**Comment by sharnoff** [comment]
status update on this: Broadly, the code seems to be working, even though it's based main from back in July. Next steps are squashing, rebasing/merging onto main and polishing it up (removing older stuff that's now unnecessary (like some fields in `agent.Runner`) and adding comments to all the new methods).

Before that, it's probably best to merge #506 — should make things on this PR simpler (even though merging/rebasing will be more complicated)

**Comment by sharnoff** [comment]
Squashed

File: pkg/agent/core/action.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+type ActionInformantDownscale struct {
+	Target api.Resources `json:"target"`
+}
+
+type ActionInformantUpscale struct {
+	Resources api.Resources `json:"resources"`
+}
+
+// Valid checks that the Action is valid. Currently, that exactly one field is set
+func (a Action) Valid() error {
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply pkg/agent/core/action.go]: No longer exists

**Comment by sharnoff** [comment]
I think this is _probably_ ready to merge, need to give it a thorough look over once more.

In its current state, this PR:

- _does not_ fix #350 or #23. Those should be easy enough to do, but leaving that for a follow-up because the behavior is kind of significantly different, and I don't want to block this PR on fixing those.
- _does_ fix #252 (if it did exist before this PR), having a couple tests to check it

Want to let this sit on staging for as long as possible, given the magnitude of the changes.

**Pull request merged**